funsor/gaussian.py

```diff
@@ -380,42 +382,74 @@ def eager_subs(self, subs):
-            batch_dim = tensors[0].dim() - 1
-            batch_shape = broadcast_shape(*(x.shape[:batch_dim] for x in tensors))
-            (loc, precision), values = tensors[:2], tensors[2:]
+        int_inputs = OrderedDict((k, d) for k, d in self.inputs.items() if d.dtype != 'real')
+        tensors = [Tensor(self.loc, int_inputs),
+                   Tensor(self.precision, int_inputs)]
+        tensors.extend(real_subs.values())
+        int_inputs, tensors = align_tensors(*tensors)
+        batch_dim = tensors[0].dim() - 1
+        batch_shape = broadcast_shape(*(x.shape[:batch_dim] for x in tensors))
+        (loc, precision), values = tensors[:2], tensors[2:]
+        offsets, event_size = _compute_offsets(self.inputs)
+        slices = [(k, slice(offset, offset + self.inputs[k].num_elements))
+                  for k, offset in offsets.items()]
+
+        # Expand all substituted values.
+        values = OrderedDict(zip(real_subs, values))
+        for k, value in values.items():
+            value = value.reshape(value.shape[:batch_dim] + (-1,))
+            if not torch._C._get_tracing_state():
+                assert value.size(-1) == self.inputs[k].num_elements
+            values[k] = value.expand(batch_shape + value.shape[-1:])
 
+        # Try to perform a complete substitution of all real variables, resulting in a Tensor.
+        if all(k in real_subs for k, d in self.inputs.items() if d.dtype == 'real'):
             # Form the concatenated value.
-            offsets, event_size = _compute_offsets(self.inputs)
             value = BlockVector(batch_shape + (event_size,))
-            for k, value_k in zip(real_subs, values):
-                offset = offsets[k]
-                value_k = value_k.reshape(value_k.shape[:batch_dim] + (-1,))
-                if not torch._C._get_tracing_state():
-                    assert value_k.size(-1) == self.inputs[k].num_elements
-                value_k = value_k.expand(batch_shape + value_k.shape[-1:])
-                value[..., offset: offset + self.inputs[k].num_elements] = value_k
+            for k, i in slices:
+                if k in values:
+                    value[..., i] = values[k]
             value = value.as_tensor()
 
             # Evaluate the non-normalized log density.
             result = -0.5 * _vmv(precision, value - loc)
-            result = Tensor(result, inputs)
+            result = Tensor(result, int_inputs)
             assert result.output == reals()
             return Subs(result, lazy_subs)
 
         # Perform a partial substution of a subset of real variables, resulting in a Joint.
-        # See "The Matrix Cookbook" (November 15, 2012) ss. 8.1.3 eq. 353.
-        # http://www.math.uwaterloo.ca/~hwolkowi/matrixcookbook.pdf
-        raise NotImplementedError('TODO implement partial substitution of real variables')
+        # We split real inputs into two sets: a for the preserved and b for the substituted.
```

Comment:
@fehiepsi if get a chance, could you check my math here? Tests are failing, but the failure could be in either linear algebra or dimension bookkeeping.